**Hamoye- Stage A Quiz**

In [1]:
import numpy as np
import pandas as pd

**Question 1**

In [2]:
#Question 1
A = [1,2,3,4,5,6] 
B = [13, 21, 34]

In [3]:
#Best way to combine them to form array A_B
A_B = A + B
print (A_B)

[1, 2, 3, 4, 5, 6, 13, 21, 34]


**Question 2**

In [4]:
#Question 2
#Answer: A
C = np.identity(3)
print (C)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


**Question 3**

In [5]:
#Question 3
fuel_ferc1 = pd.read_csv(r'C:\Users\Kolade Olawoyin\Documents\Hamoye internship\fuel_ferc1.csv')
print(fuel_ferc1)

                      record_id  utility_id_ferc1  report_year  \
0         f1_fuel_1994_12_1_0_7                 1         1994   
1        f1_fuel_1994_12_1_0_10                 1         1994   
2         f1_fuel_1994_12_2_0_1                 2         1994   
3         f1_fuel_1994_12_2_0_7                 2         1994   
4        f1_fuel_1994_12_2_0_10                 2         1994   
5        f1_fuel_1994_12_2_0_13                 2         1994   
6         f1_fuel_1994_12_2_1_1                 2         1994   
7        f1_fuel_1994_12_2_1_10                 2         1994   
8        f1_fuel_1994_12_2_2_10                 2         1994   
9         f1_fuel_1994_12_4_0_1                 4         1994   
10        f1_fuel_1994_12_4_0_2                 4         1994   
11        f1_fuel_1994_12_4_0_4                 4         1994   
12        f1_fuel_1994_12_4_0_5                 4         1994   
13        f1_fuel_1994_12_6_0_1                 6         1994   
14        

In [6]:
fuel_ferc1 = pd.DataFrame(fuel_ferc1)
fuel_ferc1

                      record_id  utility_id_ferc1  report_year  \
0         f1_fuel_1994_12_1_0_7                 1         1994   
1        f1_fuel_1994_12_1_0_10                 1         1994   
2         f1_fuel_1994_12_2_0_1                 2         1994   
3         f1_fuel_1994_12_2_0_7                 2         1994   
4        f1_fuel_1994_12_2_0_10                 2         1994   
5        f1_fuel_1994_12_2_0_13                 2         1994   
6         f1_fuel_1994_12_2_1_1                 2         1994   
7        f1_fuel_1994_12_2_1_10                 2         1994   
8        f1_fuel_1994_12_2_2_10                 2         1994   
9         f1_fuel_1994_12_4_0_1                 4         1994   
10        f1_fuel_1994_12_4_0_2                 4         1994   
11        f1_fuel_1994_12_4_0_4                 4         1994   
12        f1_fuel_1994_12_4_0_5                 4         1994   
13        f1_fuel_1994_12_6_0_1                 6         1994   
14        

In [7]:
#fuel_ferc1 in dataframe
fuel_ferc1 = pd.DataFrame(fuel_ferc1)
fuel_ferc1

record_id  utility_id_ferc1  report_year  \
0         f1_fuel_1994_12_1_0_7                 1         1994   
1        f1_fuel_1994_12_1_0_10                 1         1994   
2         f1_fuel_1994_12_2_0_1                 2         1994   
3         f1_fuel_1994_12_2_0_7                 2         1994   
4        f1_fuel_1994_12_2_0_10                 2         1994   
5        f1_fuel_1994_12_2_0_13                 2         1994   
6         f1_fuel_1994_12_2_1_1                 2         1994   
7        f1_fuel_1994_12_2_1_10                 2         1994   
8        f1_fuel_1994_12_2_2_10                 2         1994   
9         f1_fuel_1994_12_4_0_1                 4         1994   
10        f1_fuel_1994_12_4_0_2                 4         1994   
11        f1_fuel_1994_12_4_0_4                 4         1994   
12        f1_fuel_1994_12_4_0_5                 4         1994   
13        f1_fuel_1994_12_6_0_1                 6         1994   
14        f1_fuel_1994_12_6_0_2                 6         1994   
15        f1_fuel_1994_12_6_0_4                 6         1994   
16        f1_fuel_1994_12_6_0_7                 6         1994   
17       f1_fuel_1994_12_6_0_10                 6         1994   
18       f1_fuel_1994_12_6_0_13                 6         1994   
19       f1_fuel_1994_12_6_0_14                 6         1994   
20        f1_fuel_1994_12_6_1_7                 6         1994   
21        f1_fuel_1994_12_6_1_8                 6         1994   
22       f1_fuel_1994_12_6_1_10                 6         1994   
23       f1_fuel_1994_12_6_1_11                 6         1994   
24       f1_fuel_1994_12_6_1_13                 6         1994   
25       f1_fuel_1994_12_6_1_14                 6         1994   
26        f1_fuel_1994_12_7_0_1                 7         1994   
27        f1_fuel_1994_12_7_0_2                 7         1994   
28        f1_fuel_1994_12_7_0_3                 7         1994   
29        f1_fuel_1994_12_7_0_7                 7         1994   
...                         ...               ...          ...   
29493   f1_fuel_2018_12_191_0_3               191         2018   
29494   f1_fuel_2018_12_191_0_5               191         2018   
29495   f1_fuel_2018_12_191_0_6               191         2018   
29496   f1_fuel_2018_12_191_0_8               191         2018   
29497  f1_fuel_2018_12_191_0_11               191         2018   
29498   f1_fuel_2018_12_191_1_2               191         2018   
29499   f1_fuel_2018_12_191_1_3               191         2018   
29500   f1_fuel_2018_12_191_1_6               191         2018   
29501   f1_fuel_2018_12_191_1_8               191         2018   
29502   f1_fuel_2018_12_191_1_9               191         2018   
29503  f1_fuel_2018_12_191_1_14               191         2018   
29504  f1_fuel_2018_12_191_1_15               191         2018   
29505   f1_fuel_2018_12_432_0_2               432         2018   
29506   f1_fuel_2018_12_432_0_5               432         2018   
29507   f1_fuel_2018_12_432_0_8               432         2018   
29508  f1_fuel_2018_12_432_0_11               432         2018   
29509  f1_fuel_2018_12_432_0_14               432         2018   
29510   f1_fuel_2018_12_403_0_2               403         2018   
29511   f1_fuel_2018_12_403_0_6               403         2018   
29512   f1_fuel_2018_12_403_0_9               403         2018   
29513    f1_fuel_2018_12_12_0_2                12         2018   
29514    f1_fuel_2018_12_12_0_4                12         2018   
29515    f1_fuel_2018_12_12_0_7                12         2018   
29516    f1_fuel_2018_12_12_0_8                12         2018   
29517   f1_fuel_2018_12_12_0_10                12         2018   
29518   f1_fuel_2018_12_12_0_13                12         2018   
29519    f1_fuel_2018_12_12_1_1                12         2018   
29520   f1_fuel_2018_12_12_1_10                12         2018   
29521   f1_fuel_2018_12_12_1_13                12         2018

In [8]:
#Lowest average cost 
lowest_average_cost = fuel_ferc1.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].mean()

In [9]:
lowest_average_cost

fuel_type_code_pudl
coal         67.421830
gas          13.659397
nuclear    4955.157002
oil         168.877086
other        18.253856
waste        19.518122
Name: fuel_cost_per_unit_burned, dtype: float64

As we can see, Gas has the lowest with 13.659397

**Question 4**

In [10]:
#calculating the standard deviation
standard_dev = fuel_ferc1['fuel_mmbtu_per_unit'].std()
print (standard_dev)

10.600220307806886


In [11]:
#calculating the 75th Percentile
Percentile = fuel_ferc1['fuel_mmbtu_per_unit'].quantile(0.75)
print (Percentile)

17.006


**Question 5**


In [12]:
#calcuation the skewness
skew= fuel_ferc1['fuel_qty_burned'].skew()
print (skew)

15.851495469109503


In [13]:
#calcuation the kurtosis
kurt= fuel_ferc1['fuel_qty_burned'].kurtosis()
print (kurt)

651.3694501337732


**Question 6**

In [14]:
#checking for missing values
fuel_ferc1.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

fuel_unit, it has 180 missing values

In [15]:
#percent of missing values
percent_missing = fuel_ferc1.isnull().sum() * 100 / len(fuel_ferc1)

In [16]:
print(percent_missing)

record_id                       0.000000
utility_id_ferc1                0.000000
report_year                     0.000000
plant_name_ferc1                0.000000
fuel_type_code_pudl             0.000000
fuel_unit                       0.609694
fuel_qty_burned                 0.000000
fuel_mmbtu_per_unit             0.000000
fuel_cost_per_unit_burned       0.000000
fuel_cost_per_unit_delivered    0.000000
fuel_cost_per_mmbtu             0.000000
dtype: float64


fuel_unit has 0.609%

**Question 7**

Categorical and Mode computation

**Question 8**

In [17]:
#comparing correlation coefficients between fuel_cost_per_unit_burned and other columns:
# with fuel_mmbtu_per_unit
corr1 = fuel_ferc1['fuel_cost_per_unit_burned'].corr(fuel_ferc1['fuel_mmbtu_per_unit'])
print (corr1)

-0.010033808428853057


In [18]:
#comparing correlation coefficients between fuel_cost_per_unit_burned and other columns:
# with fuel_qty_burned
corr2 = fuel_ferc1['fuel_cost_per_unit_burned'].corr(fuel_ferc1['fuel_qty_burned'])
print (corr2)

-0.018535416794825225


In [19]:
#comparing correlation coefficients between fuel_cost_per_unit_burned and other columns:
# with utility_id_ferc1
corr3 = fuel_ferc1['fuel_cost_per_unit_burned'].corr(fuel_ferc1['utility_id_ferc1'])
print (corr3)

-0.037862707599972595


In [21]:
#comparing correlation coefficients between fuel_cost_per_unit_burned and other columns:
# with fuel_cost_per_unit_delivered
corr4 = fuel_ferc1['fuel_cost_per_unit_burned'].corr(fuel_ferc1['fuel_cost_per_unit_delivered'])
print (corr4)

0.011007403299165112


Answer: fuel_mmbtu_per_unit and fuel_cost_per_unit_delivered

**Question 9**

In [22]:
average_per_year = fuel_ferc1.groupby('report_year' )['fuel_cost_per_unit_burned'].mean()
average_per_year

report_year
1994    213.191996
1995    354.422555
1996     79.434995
1997    149.482132
1998    126.683519
1999    120.053605
2000    237.797111
2001    133.013010
2002    101.179365
2003    123.183919
2004    154.897398
2005    107.111221
2006    151.361249
2007    202.622082
2008    255.954415
2009    215.699318
2010    262.951203
2011    301.754519
2012    258.804138
2013    255.991023
2014    442.359068
2015    237.933197
2016    247.083959
2017    246.281796
2018    232.992036
Name: fuel_cost_per_unit_burned, dtype: float64

In [23]:
year1994 = 213.191996
year1998 = 126.683519
pct_change = ((year1998 - year1994)/year1998) *100
pct_change

-68.2870808159347

**Question 10**

In [24]:
Highest_year = fuel_ferc1.groupby('report_year')['fuel_cost_per_unit_delivered'].max()
Highest_year

report_year
1994      45115.000
1995       2396.770
1996    7793482.000
1997    7964521.000
1998     270604.000
1999       2891.100
2000    1168000.000
2001      34381.000
2002      31416.000
2003      34518.000
2004     132884.000
2005       8651.000
2006       1431.000
2007       4850.872
2008      10490.000
2009     730166.000
2010      44033.320
2011       9545.000
2012      10129.000
2013     135315.000
2014     136410.000
2015     140706.000
2016      69440.000
2017       3584.335
2018     245652.000
Name: fuel_cost_per_unit_delivered, dtype: float64

**Therefore, it is year 1997 with 7964521.000**